pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/TransactionCursorImpl.java

```diff
@@ -123,4 +204,377 @@ private void addTxnToCommittedIndex(TxnID txnID, long committedAtLedgerId) {
+
+        cursorLedger.asyncReadEntries(startEntryId, endEntryId, (rc, handle, entries, ctx) -> {
+            if (rc != BKException.Code.OK) {
+                readFuture.completeExceptionally(BKException.create(rc));
+            } else {
+                if (entries.hasMoreElements()) {
+                    List<Entry> entryList = Collections.list(entries)
+                                                       .stream()
+                                                       .map(ledgerEntry -> EntryImpl.create(ledgerEntry.getLedgerId()
+                                                           , ledgerEntry.getEntryId(), ledgerEntry.getEntry()))
+                                                       .collect(Collectors.toList());
+                    readFuture.complete(entryList);
+                } else {
+                    readFuture.completeExceptionally(new NoSuchElementException(
+                        "No more entry can read from ledger: " + handle.getId() + ", entry: " + startEntryId));
+                }
+            }
+        }, null);
+
+        return readFuture;
+    }
+
+    private CompletableFuture<List<Entry>> readEntryFromLedger(Position startSnapshotPos, ManagedLedger managedLedger) {
+        CompletableFuture<List<Entry>> readFuture = new CompletableFuture<>();
+
+        List<CompletableFuture<Void>> readAllEntryFuture = new ArrayList<>();
+        List<Entry> entryList = new ArrayList<>();
+        ManagedLedger cursorLedger = managedLedger;
+        ManagedCursor readCursor = null;
+        try {
+            readCursor = cursorLedger.newNonDurableCursor(startSnapshotPos);
+
+            while (readCursor.hasMoreEntries()) {
+                CompletableFuture<Void> readEntries = new CompletableFuture<>();
+                readCursor.asyncReadEntries(100, new AsyncCallbacks.ReadEntriesCallback() {
+                    @Override
+                    public void readEntriesComplete(List<Entry> entries, Object ctx) {
+                        synchronized (entryList) {
+                            entryList.addAll(entries);
+                        }
+                        readEntries.complete(null);
+                    }
+
+                    @Override
+                    public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
+                        readEntries.completeExceptionally(exception);
+                    }
+                }, null);
+                readAllEntryFuture.add(readEntries);
+            }
+
+            FutureUtil.waitForAll(readAllEntryFuture).whenComplete((ignore, error) -> {
+                if (error != null) {
+                    readFuture.completeExceptionally(error);
+                } else {
+                    readFuture.complete(entryList);
+                }
+            });
+
+        } catch (ManagedLedgerException e) {
+            readFuture.completeExceptionally(e);
+        } finally {
+            if (readCursor != null) {
```

Comment:
if readCursor is used in async context, you can't close it in this finally block. because when the async method is called, the finally block will be called but the read cursor is still used for reading entries.